### PR TITLE
WIP Serialize params

### PR DESCRIPTION
**Pull request by gavin-norman-sociomantic**

### CODE REVIEW

File: src/swarm/neo/client/mixins/ClientCore.d [hunk 1]
```diff
@@ -581,10 +581,10 @@ template ClientCore ( )
 
     private RequestId assign ( R, P ) ( P params )
     {
-        static assert(is(P == R.UserSpecifiedParams));
+        static assert(is(P == Const!(R.UserSpecifiedParams)));
```
Comment (mihails-strasuns-sociomantic) [inline src/swarm/neo/client/mixins/ClientCore.d]: I'd use `:` instead of `==` here, as both mutable and immutable are implicitly convertible to const.

File: src/swarm/neo/client/mixins/RequestCore.d [hunk 2]
```diff
@@ -178,7 +178,12 @@ public template RequestCore ( RequestType request_type_, ubyte request_code,
     {
         public Args args;
 
-        public SerializableReferenceType!(Notifier) notifier;
+        ubyte[Notifier.sizeof] notifier;
```
Comment (mihails-strasuns-sociomantic) [inline src/swarm/neo/client/mixins/RequestCore.d]: Not immediately obvious, why was this needed?

File: src/swarm/neo/client/mixins/RequestCore.d [hunk 3]
```diff
@@ -190,13 +195,28 @@ public template RequestCore ( RequestType request_type_, ubyte request_code,
 
     public static struct Context
     {
+        import ocean.util.serialize.contiguous.Serializer;
+        import ocean.util.serialize.contiguous.Deserializer;
+
+        private void[] user_specified_params_blob;
```
Comment (mihails-strasuns-sociomantic) [inline src/swarm/neo/client/mixins/RequestCore.d]: the host struct is reused, right?

Comment (mihails-strasuns-sociomantic) [reply src/swarm/neo/client/mixins/RequestCore.d]: maybe define it as `Contiguous!(UserSpecifiedParams)` if yes:

```D
        import ocean.util.serialize.contiguous.Util;
        
        public Contiguous!(UserSpecifiedParams) user_specified_params;

        public void serializeUserSpecifiedParams (
            Const!(UserSpecifiedParams) params )
        {
            copy(params, this.user_specified_params);
        }
```

**Review by mihails-strasuns-sociomantic — COMMENTED** [review]
Resulting code change looks pretty clean!

File: src/swarm/neo/client/mixins/RequestCore.d (hunk 2 above)
Comment (gavin-norman-sociomantic) [reply src/swarm/neo/client/mixins/RequestCore.d]: Before, setting up a struct instance looked like:
```D
Request.UserSpecifiedParams params;
...
params.notifier.set(user_notifier); // mutates params
```
If `params` is declared const, it's not possible to call `params.notifier.set(...);`.

File: src/swarm/neo/client/mixins/RequestCore.d (hunk 3 above)
Comment (david-eckardt-sociomantic) [reply src/swarm/neo/client/mixins/RequestCore.d]: We thought of that. However, `Contiguous.data` is not a public field so the de/serialiser would mess with private property. Do you think that would be better?

Comment (gavin-norman-sociomantic) [reply src/swarm/neo/client/mixins/RequestCore.d]: >the host struct is reused, right?

A `Context` instance is constructed and serialized. Each request handler is passed the serialized buffer and deserializes a `Context` struct again.

Comment (mihails-strasuns-sociomantic) [reply src/swarm/neo/client/mixins/RequestCore.d]: > However, Contiguous.data is not a public field so the de/serialiser would mess with private property

`copy` does it for you - https://github.com/sociomantic-tsunami/ocean/blob/v2.x.x/src/ocean/util/serialize/contiguous/Util.d#L68 (though `const` fix for `copy` was merged very recently and may not be available for swarm :X)

Comment (mihails-strasuns-sociomantic) [reply src/swarm/neo/client/mixins/RequestCore.d]: @gavin-norman-sociomantic so that means one GC allocation per each request or?

Comment (david-eckardt-sociomantic) [reply src/swarm/neo/client/mixins/RequestCore.d]: Sorry, I misunderstood.

Comment (gavin-norman-sociomantic) [reply src/swarm/neo/client/mixins/RequestCore.d]: >so that means one GC allocation per each request or?

Ah shit, you're right. That's no good.

**Comment by gavin-norman-sociomantic** [comment]
Updated to use shared/reused serialization buffer.

File: src/swarm/neo/client/mixins/RequestCore.d [hunk 4]
```diff
@@ -235,10 +304,10 @@ public template RequestCore ( RequestType request_type_, ubyte request_code,
 
     ***************************************************************************/
 
-    private static void notify ( ref UserSpecifiedParams params,
+    private static void notify ( UserSpecifiedParams* params,
```
Comment (david-eckardt-sociomantic) [inline src/swarm/neo/client/mixins/RequestCore.d:307]: Why this change?

Comment (gavin-norman-sociomantic) [reply src/swarm/neo/client/mixins/RequestCore.d:307]: Just for convenience. `user_params` returns a pointer to a struct now, so usage is simpler if `notify` accepts the same type.

Comment (nemanja-boric-sociomantic) [reply src/swarm/neo/client/mixins/RequestCore.d:307]: Maybe in the next major we could make this `Const!(UserSpecifiedParams)*` params? So that would make a breaking change, since the protos must update their `Notifier` definitions, to accept `Args` by const, but users would just have the read-only view of their args in the notifiers.

Comment (gavin-norman-sociomantic) [reply src/swarm/neo/client/mixins/RequestCore.d:307]: Yes, I thought of that too.

**Comment by gavin-norman-sociomantic** [comment]
Closing. To be replaced with a proper version.
